File: 2416.字符串的前缀分数和/2416-字符串的前缀分数和.js
/**
 * @param {string[]} words
 * @return {number[]}
 */
class Node{
    constructor(){
        this.son=new Map();
        this.ids=[];
        this.score=0;
    }
}

var sumPrefixScores = function(words) {
    const n=words.length;
    let root=new Node();
    words.forEach((word, index)=>{
        let cur=root;
        for(let ch of word){
            let child;
            if(cur.son.has(ch)){
                child=cur.son.get(ch);
            }else{
                child=new Node;
                cur.son.set(ch, child);
            }
            cur=child;
            cur.score++;
        }
        cur.ids.push(index);
    });
    let ans=new Array(n).fill(0);

    function dfs(node, sum){
        sum+=node.score;
        for(let i of node.ids){
            ans[i]=sum;
        }
        for(let child of node.son.values()){
            dfs(child, sum);
        }
    }

    dfs(root, 0);
    return ans;
};

File: 2416.字符串的前缀分数和/2416-字符串的前缀分数和.py
Trie = lambda : defaultdict(Trie)
CNT = '#'

class Solution:
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        tr = Trie()
        for w in words:
            cur = tr
            for ch in w:
                cur = cur[ch]
                cur[CNT] = cur.get(CNT, 0) + 1
        
        res = []
        for w in words:
            cur, ans = tr, 0
            for ch in w:
                cur = cur[ch]
                ans += cur[CNT]
            res.append(ans)
        return res